File: src/main/java/me/lokka30/levelledmobs/customdrops/CustomDropBase.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.customdrops;

import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;
import me.lokka30.levelledmobs.misc.CachedModalList;
import me.lokka30.levelledmobs.util.Utils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Base class for all custom drops including custom commands
 *
 * @author stumper66
 * @since 3.0.0
 */
public abstract class CustomDropBase implements Cloneable {

    CustomDropBase(@NotNull final CustomDropsDefaults defaults) {
        this.excludedMobs = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
        this.amount = defaults.amount;
        this.permissions = new LinkedList<>();
        this.playeerVariableMatches = new LinkedList<>();
    }

    private int amount;
    int amountRangeMin;
    private int amountRangeMax;
    private boolean hasAmountRange;
    public int minLevel;
    public int maxLevel;
    public int priority;
    int maxDropGroup;
    int minPlayerLevel;
    int maxPlayerLevel;
    public boolean useChunkKillMax;
    public float chance;
    boolean playerCausedOnly;
    boolean noSpawner;
    String groupId;
    String playerLevelVariable;
    final public @NotNull List<String> permissions;
    final public @NotNull List<String> playeerVariableMatches;
    final @NotNull Set<String> excludedMobs;
    CachedModalList<DeathCause> causeOfDeathReqs;

    public int getAmount() {
        return this.amount;
    }

    public void setAmount(final int amount) {
        this.amount = amount;
        if (this.amount > 64) {
            this.amount = 64;
        }
        if (this.amount < 1) {
            this.amount = 1;
        }
        this.hasAmountRange = false;
    }

    int getAmountRangeMin() {
        return this.amountRangeMin;
    }

    int getAmountRangeMax() {
        return this.amountRangeMax;
    }

    boolean getHasAmountRange() {
        return this.hasAmountRange;
    }

    @NotNull String getAmountAsString() {
        if (this.hasAmountRange) {
            return String.format("%s-%s", this.amountRangeMin, this.amountRangeMax);
        } else {
            return String.valueOf(this.amount);
        }
    }

    boolean setAmountRangeFromString(final @Nullable String numberOrNumberRange) {
        if (numberOrNumberRange == null || numberOrNumberRange.isEmpty()) {
            return false;
        }

        if (!numberOrNumberRange.contains("-")) {
            if (!Utils.isInteger(numberOrNumberRange)) {
                return false;
            }

            this.amount = Integer.parseInt(numberOrNumberRange);
            this.hasAmountRange = false;
            return true;
        }

        final String[] nums = numberOrNumberRange.split("-");
        if (nums.length != 2) {
            return false;
        }

        if (!Utils.isInteger(nums[0].trim()) || !Utils.isInteger(nums[1].trim())) {
            return false;
        }
        this.amountRangeMin = Integer.parseInt(nums[0].trim());
        this.amountRangeMax = Integer.parseInt(nums[1].trim());
        this.hasAmountRange = true;

        return true;
    }

    @SuppressWarnings("unchecked")
    public CustomDropBase cloneItem() {
        CustomDropBase copy = null;
        try {
            copy = (CustomDropBase) super.clone();
            if (this.causeOfDeathReqs != null) {
                copy.causeOfDeathReqs = (CachedModalList<DeathCause>) this.causeOfDeathReqs.clone();
            }
        } catch (final Exception e) {
            e.printStackTrace();
        }

        return copy;
    }
}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/plugin/bukkit/customdrops/CustomDropBase.kt
package io.github.arcaneplugins.levelledmobs.plugin.bukkit.customdrops

import io.github.arcaneplugins.levelledmobs.plugin.bukkit.misc.CachedModalList
import io.github.arcaneplugins.levelledmobs.plugin.bukkit.misc.Utils
import java.util.LinkedList
import java.util.TreeSet

abstract class CustomDropBase : Cloneable {
    var amountRangeMin = 0
        private set
    var amountRangeMax = 0
        private set
    var hasAmountRange = false
        private set
    var minLevel = 0
    var maxLevel = 0
    var priority = 0
    var maxDropGroup = 0
    var minPlayerLevel = 0
    var maxPlayerLevel = 0
    var useChunkKillMax = false
    var chance = 0f
    var playerCausedOnly = false
    var noSpawner = false
    var groupId: String? = null
    var playerLevelVariable: String? = null
    val permissions: List<String>
    val playerVariableMatches: List<String>
    val excludedMobs: Set<String>
    var causeOfDeathReqs: CachedModalList<DeathCause>? = null

    init {
        excludedMobs = TreeSet(String.CASE_INSENSITIVE_ORDER)
        permissions = LinkedList()
        playerVariableMatches = LinkedList()
    }

    var amount: Int = 0
        set(value) {
            field = 1.coerceAtLeast(value)
            field = 64.coerceAtMost(value)
        }

    fun getAmountAsString() : String{
        return if (this.hasAmountRange){
            String.format("%s-%s", amountRangeMin, amountRangeMax)
        } else{
            this.amount.toString()
        }
    }

    fun setAmountRangeFromString(numberOrNumberRange: String?): Boolean{
        if (numberOrNumberRange.isNullOrEmpty()) {
            return false
        }

        if (!numberOrNumberRange.contains("-")) {
            if (!Utils.isInteger(numberOrNumberRange)) {
                return false
            }
            amount = numberOrNumberRange.toInt()
            hasAmountRange = false
            return true
        }

        val nums = numberOrNumberRange.split("-")
        if (nums.size != 2) return false

        if (!Utils.isInteger(nums[0].trim()) || !Utils.isInteger(nums[1].trim())){
            return false
        }

        this.amountRangeMin = nums[0].trim().toInt()
        this.amountRangeMax = nums[1].trim().toInt()
        this.hasAmountRange = true

        return true
    }

    @Suppress("UNCHECKED_CAST")
    override fun clone(): CustomDropBase{
        val copy: CustomDropBase = super.clone() as CustomDropBase
        if (this.causeOfDeathReqs != null){
            copy.causeOfDeathReqs = this.causeOfDeathReqs!!.clone() as CachedModalList<DeathCause>
        }

        return copy
    }
}